### PR TITLE
Initial SM4 key search implementation

### PR DESCRIPTION
**Pull request by sylvainpelissier**
- [x] Mark this if you consider it ready to merge
- [x] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**
Initial SM4 key search implementation in `/ca` command. Add SM4 key search test. Refactor cmd_search.c to have common code for key search.

### CODE REVIEW

File: libr/include/r_crypto/r_sm4.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* Family Key FK */
+static const ut32 FK[4] = { 0xa3b1bac6, 0x56aa3350, 0x677d9197, 0xb27022dc };
+
+/* Constant Key CK */
+const ut32 CK[32] = {
```
Comment (trufae) [inline libr/include/r_crypto/r_sm4.h:14]:
```suggestion
static const ut32 CK[32] = {
```

File: libr/include/r_crypto/r_sm4.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+#define R_SM4_H
+
+#include <r_util.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* Family Key FK */
+static const ut32 FK[4] = { 0xa3b1bac6, 0x56aa3350, 0x677d9197, 0xb27022dc };
```
Comment (trufae) [inline libr/include/r_crypto/r_sm4.h:11]: better move these into the .c

File: libr/include/r_crypto/r_sm4.h (hunk 1 above)
Comment (trufae) [reply libr/include/r_crypto/r_sm4.h:14]: and move into the .c

Comment (sylvainpelissier) [reply libr/include/r_crypto/r_sm4.h]: I agree for FK values but for CK values I also need them in the key search so I need themin the h file.

**Comment by trufae** [comment]
clang-11: warning: argument unused during compilation: '-pie' [-Wunused-command-line-argument]
[5232](https://github.com/radareorg/radare2/actions/runs/3037423609/jobs/4889842473#step:3:5233)
wasm-ld: error: /home/runner/work/radare2/radare2/libr/search/libr_search.a(sm4_find.o): undefined symbol: sm4_RK
[5233](https://github.com/radareorg/radare2/actions/runs/3037423609/jobs/4889842473#step:3:5234)
wasm-ld: error: /home/runner/work/radare2/radare2/libr/search/libr_search.a(sm4_find.o): undefined symbol: sm4_RK
[5234](https://github.com/radareorg/radare2/actions/runs/3037423609/jobs/4889842473#step:3:5235)
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)

File: libr/include/r_crypto/r_sm4.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+#define R_SM4_H
+
+#include <r_util.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* Constant Key CK */
+static const ut32 sm4_CK[32] = {
```
Comment (trufae) [inline libr/include/r_crypto/r_sm4.h:11]: move it into the C, thres no need to expose those constants globally

Comment (sylvainpelissier) [reply libr/include/r_crypto/r_sm4.h:11]: I need them in `libr/crypto/p/crypto_sm4.c`

**Comment by trufae** [comment]
I will take over this, dont worry :) thanks for the contrib!

**Pull request closed without merging**